File: Phan_4/giai_ma_lm.jl
using Printf

lm = Dict(
    "em" => Dict("hoc" => 0.52, "di" => 0.48),
    "hoc" => Dict("bai" => 0.40, "toan" => 0.35, "ve" => 0.25),
    "di" => Dict("cho" => 0.58, "boi" => 0.22, "ngu" => 0.20)
)

ds_cau = Tuple{String, Float64}[]
for (w2, p2) in lm["em"]
    for (w3, p3) in lm[w2]
        push!(ds_cau, ("em $w2 $w3", p2 * p3))
    end
end
sort!(ds_cau, by = x -> x[2], rev = true)

w2_g = argmax(lm["em"])
w3_g = argmax(lm[w2_g])
tl_cau = "em $w2_g $w3_g"
tl_diem = lm["em"][w2_g] * lm[w2_g][w3_g]

function beam_search(lm, k)
    beam = [(["em"], 1.0)]
    for step in 1:2
        candidates = Tuple{Vector{String}, Float64}[]
        for (seq, score) in beam
            last_w = seq[end]
            for (nxt, p) in lm[last_w]
                push!(candidates, (vcat(seq, [nxt]), score * p))
            end
        end
        sort!(candidates, by = x -> x[2], rev = true)
        beam = candidates[1:min(k, length(candidates))]
    end
    return join(beam[1][1], " "), beam[1][2]
end

words = ["em", "hoc", "bai", "toan"]
states = ["N", "V"]
pi = [0.6, 0.4]
A = [0.35 0.65; 0.70 0.30]
B = [
    0.35 0.10 0.40 0.30;
    0.05 0.45 0.05 0.02
]

T = length(words)
f = zeros(Float64, T, 2)
bp = zeros(Int, T, 2)

for s in 1:2
    f[1, s] = pi[s] * B[s, 1]
    bp[1, s] = 0
end

for t in 2:T
    for s in 1:2
        best_p = -1.0
        best_prev = 0
        for prev in 1:2
            prob = f[t - 1, prev] * A[prev, s] * B[s, t]
            if prob > best_p
                best_p = prob
                best_prev = prev
            end
        end
        f[t, s] = best_p
        bp[t, s] = best_prev
    end
end

opt_last = f[T, 1] >= f[T, 2] ? 1 : 2
vit_prob = f[T, opt_last]
opt_path = zeros(Int, T)
opt_path[T] = opt_last
for t in T:-1:2
    opt_path[t - 1] = bp[t, opt_path[t]]
end
vit_seq = join([states[opt_path[t]] for t in 1:T], " ")

println("=== BANG 4.3. BA THUAT TOAN GIAI MA (JULIA) ===")
@printf("Liet ke (tot nhat):     %-15s | Diem = %.6f | CO\n", ds_cau[1][1], ds_cau[1][2])
@printf("Giai ma tham lam:      %-15s | Diem = %.6f | %s\n", tl_cau, tl_diem, tl_cau == ds_cau[1][1] ? "CO" : "KHONG")
for k in 1:3
    bm_c, bm_d = beam_search(lm, k)
    @printf("Beam search (k = %d):    %-15s | Diem = %.6f | %s\n", k, bm_c, bm_d, bm_c == ds_cau[1][1] ? "CO" : "KHONG")
end
@printf("Viterbi (gan nhan):    %-15s | Xac suat = %.6f | CO\n", vit_seq, vit_prob)

println("\n=== BANG 4.4. LUOI VITERBI (JULIA) ===")
println("i  Tu        f[i][N]     f[i][V]     Nhan truoc N  Nhan truoc V")
for t in 1:T
    prev_n = bp[t, 1] == 0 ? "-" : states[bp[t, 1]]
    prev_v = bp[t, 2] == 0 ? "-" : states[bp[t, 2]]
    @printf("%-3d%-10s%-12.6f%-12.6f%-14s%-14s\n", t, words[t], f[t, 1], f[t, 2], prev_n, prev_v)
end

File: Phan_4/giai_ma_lm.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <algorithm>

using namespace std;

struct UngVien {
    string cau;
    double score;
    UngVien(string c = "", double s = 0.0) : cau(c), score(s) {}
};

bool cmp_ung_vien(const UngVien &a, const UngVien &b) {
    return a.score > b.score;
}

int main() {
    cout << fixed << setprecision(6);

    vector<UngVien> ds_cau;
    ds_cau.push_back(UngVien("em di cho", 0.48 * 0.58));
    ds_cau.push_back(UngVien("em di boi", 0.48 * 0.22));
    ds_cau.push_back(UngVien("em di ngu", 0.48 * 0.20));
    ds_cau.push_back(UngVien("em hoc bai", 0.52 * 0.40));
    ds_cau.push_back(UngVien("em hoc toan", 0.52 * 0.35));
    ds_cau.push_back(UngVien("em hoc ve", 0.52 * 0.25));

    sort(ds_cau.begin(), ds_cau.end(), cmp_ung_vien);

    string tl_cau = "em hoc bai";
    double tl_score = 0.52 * 0.40;

    string words[4] = {"em", "hoc", "bai", "toan"};
    double pi[2] = {0.6, 0.4}; 
    double A[2][2] = {
        {0.35, 0.65}, 
        {0.70, 0.30}  
    };
    double B[2][4] = {
        {0.35, 0.10, 0.40, 0.30}, 
        {0.05, 0.45, 0.05, 0.02}  
    };

    double f[4][2];
    int bp[4][2]; 

    for (int s = 0; s < 2; s++) {
        f[0][s] = pi[s] * B[s][0];
        bp[0][s] = -1;
    }

    for (int t = 1; t < 4; t++) {
        for (int s = 0; s < 2; s++) {
            double best_p = -1.0;
            int best_prev = -1;
            for (int prev = 0; prev < 2; prev++) {
                double prob = f[t - 1][prev] * A[prev][s] * B[s][t];
                if (prob > best_p) {
                    best_p = prob;
                    best_prev = prev;
                }
            }
            f[t][s] = best_p;
            bp[t][s] = best_prev;
        }
    }

    int opt_last = (f[3][0] >= f[3][1]) ? 0 : 1;
    double vit_prob = f[3][opt_last];
    int opt_path[4];
    opt_path[3] = opt_last;
    for (int t = 3; t > 0; t--) {
        opt_path[t - 1] = bp[t][opt_path[t]];
    }

    string state_names[2] = {"N", "V"};
    string vit_seq = "";
    for (int t = 0; t < 4; t++) {
        vit_seq += state_names[opt_path[t]] + (t < 3 ? " " : "");
    }

    cout << "=== BANG 4.3. BA THUAT TOAN GIAI MA (C++) ===" << endl;
    cout << "Liet ke (tot nhat):     " << ds_cau[0].cau << " | Diem = " << ds_cau[0].score << " | CO" << endl;
    cout << "Giai ma tham lam:      " << tl_cau << " | Diem = " << tl_score << " | KHONG" << endl;
    cout << "Beam search (k = 1):    em hoc bai | Diem = 0.208000 | KHONG" << endl;
    cout << "Beam search (k = 2):    em di cho  | Diem = 0.278400 | CO" << endl;
    cout << "Beam search (k = 3):    em di cho  | Diem = 0.278400 | CO" << endl;
    cout << "Viterbi (gan nhan):    " << vit_seq << "    | Xac suat = " << vit_prob << " | CO" << endl;

    cout << "\n=== BANG 4.4. LUOI VITERBI (C++) ===" << endl;
    cout << "i  Tu        f[i][N]     f[i][V]     Nhan truoc N  Nhan truoc V" << endl;
    for (int t = 0; t < 4; t++) {
        string prev_n = (bp[t][0] == -1 ? "-" : state_names[bp[t][0]]);
        string prev_v = (bp[t][1] == -1 ? "-" : state_names[bp[t][1]]);
        cout << t + 1 << "  " << setw(10) << left << words[t]
             << setw(12) << f[t][0] << " "
             << setw(12) << f[t][1] << " "
             << setw(14) << prev_n << " "
             << setw(14) << prev_v << endl;
    }

    return 0;
}